[Workers] Update Fauna tutorial (#10050)
cloudflare/cloudflare-docs: content/workers/tutorials/store-data-with-fauna/index.md

- **weightLbs** - A floating point number that represents the weight in pounds of the product.
- **quantity** A non-negative integer that represents how many items of a particular product there are in the inventory.

> Documents are stored in the **Products** [collection](https://docs.fauna.com/fauna/current/reference/schema_entities/collection/). Collections in document databases are groups of related documents.

Author: You are right. I'll fix it.

type Bindings = {
  FAUNA_SECRET: string;
};

type Variables = {
  faunaClient: Client;
};

type Product = {
  id: string;
  serialNumber: number;
  title: string;
  weightLbs: number;
  quantity: number;
};

const app = new Hono<{ Bindings: Bindings; Variables: Variables }>();

app.use('*', async (c, next) => {
  const faunaClient = new Client({
    secret: c.env.FAUNA_SECRET,
  });
  c.set('faunaClient', faunaClient);
  await next();
});

app.get('/', (c) => {
  return c.text('Hello World');
});

export default app;
```

This is custom middleware to initialize the Fauna client and set the instance with `c.set()` for later use in another handler:

```js
---
header: Custom middleware for the Fauna Client
---
app.use('*', async (c, next) => {
  const faunaClient = new Client({
    secret: c.env.FAUNA_SECRET,
  });
  c.set('faunaClient', faunaClient);
  await next();
});
```

You can access `FAUNA_SECRET` environment variable from `c.env.FAUNA_SECRET`. Workers run on a [custom JavaScript runtime](https://developers.cloudflare.com/workers/runtime-apis/nodejs/) instead of Node.js, so you cannot use `process.env` to access your environment variables.

### Create product documents

> Add your first Hono handler to the `src/index.ts` file. This route accepts `POST` requests to the `/products` endpoint:

Reviewer: Does adding a Hono handler create product documents?

Author: The handler `POST /products` which is described here creates product documents.

Author: Perhaps the naming is unclear. We call this a `handler` and also a `route`.

```ts
---
header: Create product documents
---
app.post('/products', async (c) => {
  const { serialNumber, title, weightLbs } = await c.req.json<Omit<Product, 'id'>>();
  const query = fql`Products.create({
    serialNumber: ${serialNumber},
    title: ${title},
    weightLbs: ${weightLbs},
    quantity: 0
  })`;
  const result = await c.get('faunaClient').query<Product>(query);
  return c.json(result.data);
});

```

{{<Aside type="warning" header="Handler order">}}

In Hono, you should place your handler below the custom middleware.
This is because middleware and handlers are executed in sequence from top to bottom.
If you place the handler first, you cannot retrieve the instance of the Fauna client using `c.get('faunaClient')`.

{{</Aside>}}

> This route applied an FQL query in the `fql` function that creates a new document in the **Product** collection:

Author: Yes. Should be `Products`.

```js
---
header: Create query in FQL inside JavaScript
---
fql`Products.create({
  serialNumber: ${serialNumber},
  title: ${title},
  weightLbs: ${weightLbs},
  quantity: 0
})`
```

> To review what a document looks like, run the following query in the Fauna dashboard > Explorer > Region name > Database name like a `cloudflare_rest_api` > the **SHELL** window:

Reviewer: Is this asking the user to create a new database? This step is unclear.

### Base inventory logic

Replace the contents of your `src/worker.ts` file with the skeleton of your API:

```ts
---
header: src/worker.ts (skeleton)
---
import { Hono } from 'hono';
import faunadb from 'faunadb';
import { getFaunaError } from './utils';

const { Create, Collection, Get, Ref, Delete, Add, Select, Let, Var, Update } =
  faunadb.query;

type Bindings = {
  FAUNA_SECRET: string;
};

type Variables = {
  faunaClient: faunadb.Client;
};

const app = new Hono<{ Bindings: Bindings; Variables: Variables }>();

app.use('*', async (c, next) => {
  const faunaClient = new faunadb.Client({
    secret: c.env.FAUNA_SECRET,
  });
  c.set('faunaClient', faunaClient);
  await next();
});

app.get('/', (c) => {
  return c.text('Hello World!');
});

export default app;

```

This is a custom middleware to initialize the Fauna client and set the instance with `c.set()` for later use in another handler:

```js
---
header: Custom middleware for the Fauna Client
---
app.use('*', async (c, next) => {
 const faunaClient = new faunadb.Client({
  secret: c.env.FAUNA_SECRET,
 });
 c.set('faunaClient', faunaClient);
 await next();
});
```

> You can access `FAUNA_SECRET` environment variables from `c.env.FAUNA_SECRET`. Workers run on a custom JavaScript runtime instead of Node.js, so you can't use `process.env` to access your environment variables.

Reviewer: ```suggestion
You can access `FAUNA_SECRET` environment variable from `c.env.FAUNA_SECRET`. Workers run on a [custom JavaScript runtime](https://developers.cloudflare.com/workers/runtime-apis/nodejs/) instead of Node.js, so you cannot use `process.env` to access your environment variables.
```

```ts
---
header: Return the document in the response body
---
return c.json(result);
```

If not, an error is returned.

### Delete product documents

The logic to delete product documents is similar to the logic for retrieving products. Add the following route to your `src/worker.ts` file:

```ts
---
header: Delete product documents
---
app.delete('/products/:productId', async (c) => {
  const productId = c.req.param('productId');
  const result = await c
    .get('faunaClient')
    .query(Delete(Ref(Collection('Products'), productId)));
  return c.json(result);
});
```

The only difference with the previous route is that you use the [Delete][fql-delete] function to delete a document by providing its reference.

When the delete operation is successful, Fauna returns the deleted document and the route forwards the deleted document in the response's body. If not, an error is returned.

## Test and deploy your Worker

Before deploying your Worker, test it locally by using Wrangler's [dev][wrangler-dev] command.
You can execute it with NPM or Yarn:

{{<tabs labels="NPM | Yarn">}}
{{<tab label="npm" >}}

```sh
---
header: Deploy your Worker
---

> $ npm run start

Reviewer: This is saying to use Wrangler dev, but the header is deploy Worker and the command is not dev or deploy?

Author: The header should be "Develop", not "Deploy".

```

{{</tab>}}
{{<tab label="yarn" >}}

```sh
---
header: Deploy your Worker
---

> $ yarn start

Reviewer: same comment as above

}
```

Finally, deploy your Worker using the [`wrangler deploy`][wrangler-publish] command.
You can execute this with NPM or Yarn:

{{<tabs labels="NPM | Yarn">}}
{{<tab label="npm" >}}

```sh
---
header: Deploy your Worker
---

> $ npm run deploy

Reviewer: ```suggestion
$ npx wrangler deploy
```

}
```

Update your Worker by deploying it to Cloudflare.

{{<tabs labels="NPM | Yarn">}}
{{<tab label="npm" >}}

```sh
---
header: Update your Worker in Cloudflare
---
$ npm run deploy
```

{{</tab>}}
{{<tab label="yarn" >}}

```sh
---
header: Update your Worker in Cloudflare
---
$ yarn deploy
```

{{</tab>}}
{{</tabs>}}

## Complete code

If you would like to review the full source code for this application, you can find the repository [on GitHub](https://github.com/yusukebe/fauna-workers).

## Clean up

> To remove the resources you create in this tutorial, delete your Worker in the Cloudflare dashboard > **Workers & Pages** > select your Worker > **Manage Service** > **Delete**:

Reviewer: ```suggestion
To remove the resources you create in this tutorial, delete your Worker in the Cloudflare dashboard > **Workers & Pages** > select your Worker > **Manage application** > **Delete**:
```

[fql-select]: https://docs.fauna.com/fauna/current/api/fql/functions/select?lang=shell&utm_source=Cloudflare&utm_medium=referral&utm_campaign=Q4_CF_2021
[fql-update]: https://docs.fauna.com/fauna/current/api/fql/functions/update?lang=shell&utm_source=Cloudflare&utm_medium=referral&utm_campaign=Q4_CF_2021
[fql-var]: https://docs.fauna.com/fauna/current/api/fql/functions/var?lang=shell&utm_source=Cloudflare&utm_medium=referral&utm_campaign=Q4_CF_2021
[hono]: https://hono.dev/
[http-status-codes]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
[wrangler-dev]: /workers/wrangler/commands/#dev

> [wrangler-publish]: /workers/wrangler/commands/#publish

Reviewer: ```suggestion
[wrangler-deploy]: /workers/wrangler/commands/#deploy
```
